### ISSUE TITLE
SVGStringParser.cpp -- a sample SVG path that SVG Native interprets differently from other SVG viewers

### ISSUE THREAD

**Issue by dejanyy** [adobe/svg-native-viewer#117]
Attached is a sample and stripped down SVG file that SVG Native seems to interpret (and render) differently from all other SVG viewers I tired (e.g. Firefox, Chrome, Safari, Apple's Preview). All these SVG renders show the attached SVG file as a glyph "K". But with SVG Native, the top diagonal line in K is distorted.

I've analyzed the code in SVGStringParser.cpp (around lines 385 - 456) and, it appears to me, that something is not quite right with path processing logic in the ParsePathString function. Maybe I don't fully understand this code, however, it seems a bit strange to me how the startSet boolean variable is used.

I was able to make some very small changes to this code (i.e. modified just a few lines) to basically remove dependence on the startSet variable. With these changes made, the attached SVG file renders exactly the same as with other SVG viewers I tested.

If you are able to confirm that SVG Native is not interpreting this particular SVG file correctly, would you be interested in receiving and reviewing my changes?

[strange_looking_k.svg.txt](https://github.com/adobe/svg-native-viewer/files/5132331/strange_looking_k.svg.txt)

**Comment by dejanyy** [comment]
One more comment. In the above SVG file, the glyph id is associated with the root <svg> element:

<svg id="glyph53" ...
<g><path ... ></g>
</svg>

I believe this is contrary to the OpenType SVG spec (judging by the "Glyph Identifiers" section at https://docs.microsoft.com/en-us/typography/opentype/spec/svg), but who knows...

It looks like SVG Native can't find glyph ids that are attached to the root svg element.

However, the above SVG file is extracted from a font called BungeeColor-Regular_svg.ttf, so it appears there are fonts out there that like to attach glyph ids to the root svg element.

**Comment by dirkschulze** [comment]
@dejanyy Thanks a lot for reporting this inconsistency. Regardless of the spec, in this project we generally try to align with browser implementations. If all browsers are rendering it consistently, then the current implementation of svg-native-viewer should change.

I'd love to review a patch. Maybe you could elaborate what is incorrect on the use of `startSet`?

Please open another issue for the ID of the root element. I am currently not sure if that ID should match or not but would check.

**Comment by dejanyy** [comment]
Attached is my modified code snippet that reflects lines 385 - 456 of SVGStringParser.cpp. My changes are indicated with /* CHANGED */ (9 instances in total). Basically, I removed the logic that was based on the startSet boolean variable and, instead, just tried to make it work using startX and startY (and to avoid any chance of UB, it was necessary to initialize these variables at the beginning). I changed the code this way because, in my experiments with other browsers, I concluded they don't rely on something like the startSet condition. Also, I could not understand what startSet really does; the logic seemed strange to me. I don't know if there is something in the SVG specification that would require startSet to be present (and, to be honest, I don't care, I just wanted to see if I can get the same result as with other browsers with minimal code changes). So my code is below, feel free to review it and decide how you'd like to proceed. As for my second report related to the root SVG element, sure, I will submit another issue.

[changes.txt](https://github.com/adobe/svg-native-viewer/files/5491322/changes.txt)

**Comment by dirkschulze** [comment]
Ah! Thanks @dejanyy! The code path is supposed to save the starting point of the previous subpath.

Short example:
```xml
<svg xmlns="http://www.w3.org/2000/svg" width="100" height="100" viewBox="0 0 100 100">
<path d="M 0 0 L 50 0 L 100 0 z L 100 100 L 0 100 z" fill="green"/>
</svg>
```
After the first subpath (`M 0 0 L 50 0 L 100 0 z`) gets closed, the starting point is supposed to be `0 0`. So the starting point of the first point of the  previous subpath. Since subpath need to start with an `L`/`l` or `M`/`m`, this flag checking is only for moveTo, lineTo and close.

However, the logic indeed is not working properly it seems. And that might be the issue you see in your example. I'll debug it further.

**Comment by dirkschulze** [comment]
Fixed by #124
